Melhoria na estrutura docker de desenvolvimento

msoares94 opened the pull request:
**DESCRIÇÃO:**

Melhoria na estrutura docker de desenvolvimento!

Foi adicionado configurações no NGinx para direcionamento das requisições na porta 80 para 443, devido a isso foi adicionado um diretório ssl com o certificado e chave auto-assinada para o host localhost

Foi reordenado e organizado o docker-compose com intuito de deixar os servicos de banco de dados em primeira ordem de apresentação e depois em sequência

Foi restruturado o diretório de volume para deixar os database postgresql e redis na mesma hierarquia e foi adicionado no .gitignore os novos diretórios de volumes

Foi adicionado a variável COMPOSER_VERSION para a imagem do PHP

**AMBIENTE:**

- Plataforma utilizada (Docker)
- Sistema operacional e versão (Ubuntu 20.04)

msoares94 committed:
Melhorando estrutura docker de desenvolvimento, separando melhor os diretórios e adicionando arquivos auxiliares para uma instalação mais completa para desenvolvimento.

msoares94 committed:
Ajustando comando de COPY para copiar todo o conteúdo

msoares94 committed:
Ajustando o arquivo de configuração do i-Educar para não trabalhar atualmente com SSL para desenvolvimento

msoares94 committed:
Atualizando chave para host localhost

msoares94 committed:
Atualizando conteúdo e nome

msoares94 committed:
Adicionando snippets

msoares94 committed:
Removendo arquivo de extensão errada

msoares94 committed:
Ajustando conteúdo da chave para representar localhost

msoares94 committed:
Merge branch 'docker' of https://github.com/marcosoliveirasoares94/i-educar into docker

msoares94 committed:
Corrigindo duplicação de chaves devido a erro de merge

edersoares commented on docker/php/8.1/fpm-alpine/locale:
```diff
@@ -0,0 +1,3 @@
+#  File generated by update-locale
+LANG="pt_BR.UTF-8"
+LANGUAGE="pt_BR:pt:en"
```
Não tenho certeza, mas acredito o locale não deve ser alterado devido a questões de ponto flutuante no cálculo da nota. 

@tiagocamargo talvez possa ajudar nesta questão.

msoares94 replied:
Ok, aguardo o @tiagocamargo passar um retorno, com base na resposta dele ajusto a alteração.

msoares94 replied:
> Ok, aguardo o @tiagocamargo passar um retorno, com base na resposta dele ajusto a alteração.

UP!

tiagocamargo replied:
Atualmente utilizamos `en_US.UTF-8`. Desconheço os impactos de utilizar `pt_BR`. 
Mas além do ponto flutuante do PHP, pode haver problemas com postgres, que utiliza também o `en_US` para formatar número.

msoares94 replied:
> Atualmente utilizamos `en_US.UTF-8`. Desconheço os impactos de utilizar `pt_BR`. Mas além do ponto flutuante do PHP, pode haver problemas com postgres, que utiliza também o `en_US` para formatar número.

@tiagocamargo, entendido, vou manter o en_US.UTF-8

@jrbarros, irei enviar mais essa alteração, ok?

edersoares commented on docker/php/8.1/fpm-alpine/Dockerfile:
```diff
@@ -50,4 +51,5 @@ RUN apk del .phpize_deps
 RUN ln -s /var/www/ieducar/artisan /usr/local/bin/artisan
 
-RUN curl -sS https://getcomposer.org/installer | php -- --install-dir=/usr/local/bin --filename=composer
+ENV COMPOSER_VERSION=2.5.1
```
Boa! Consegue mover para o início do arquivo por favor, facilita quem procura quais variáveis podem ser modificadas.

msoares94 replied:
Feito @edersoares!

Tinha deixado no final devido ser uma variável que pode ser modificada e nesse caso a compilação poderia ser mais rápido devido o Docker utilizar as camadas anteriores não alteradas.

edersoares reviewed (commented):
@marcosoliveirasoares94 obrigado pela contribuição.

A gente entende que arquivos em branco muitas vezes dificultam o entendimento de pessoas não tão técnicas, então sugiro mantermos o mais simples possível. Porém compreendi bem a sua contribuição e acredito que poderíamos fazer uma documentação e linkar no README do projeto, o que acha?

msoares94 committed:
Movendo variável.

Movendo variável que define a versão do composer para o início do arquivo.

msoares94 committed:
Melhorando a organização do arquivo como também organizando os volumes de dados dentro de data

msoares94 committed:
Adicionando diretório de volume ao gitignote

msoares94 committed:
Renomeando arquivo e separando as configurações de SSL par ao arquivo ssl-params.conf

msoares94 committed:
Adicionando conteúdo a arquivo vazio voltado para melhoria para o desenvolvedor

jrbarros commented:
@marcosoliveirasoares94 muito obrigado pela contribuição, estou validando as alterações e gostaria de sugerir:
- Na pasta pg-dump, deixar  ela apenas com o .gitignore, assim ajuda no dia a dia, evitando fazer o commit da base inicial
- Criar a documentação em um novo readme apena para o docker, para facilitar o entendimento e as possíilidades de configuração.

Melhoria:
Com essa ambla melhoria, restou apenas a possibilidade de alterar o postgresql.conf na imagem do postgres, consegue ver isso?

marcosoliveirasoares94 commented:
> @marcosoliveirasoares94 muito obrigado pela contribuição, estou validando as alterações e gostaria de sugerir:
> 
> * Na pasta pg-dump, deixar  ela apenas com o .gitignore, assim ajuda no dia a dia, evitando fazer o commit da base inicial
> * Criar a documentação em um novo readme apena para o docker, para facilitar o entendimento e as possíilidades de configuração.
> 
> Melhoria: Com essa ambla melhoria, restou apenas a possibilidade de alterar o postgresql.conf na imagem do postgres, consegue ver isso?

Olá @jrbarros, obrigado pelo retorno!
Consigo ver sim, em breve realizarei os ajustes!

msoares94 committed:
Delete initial-dump.sql

msoares94 committed:
Na pasta pg-dump, deixando ela apenas com o .gitignore, assim ajuda no dia a dia, evitando fazer o commit da base inicial.

msoares94 committed:
Criando a documentação em um novo readme apena para o docker, para facilitar o entendimento e as possíilidades de configuração.

msoares94 committed:
Ajustando o Readme padrão pois foi extraído a parte de instalação via Docker para outro Readme específico

msoares94 committed:
Possibilitando alterar o postgresql.conf na imagem do postgres.

marcosoliveirasoares94 commented:
> @marcosoliveirasoares94 muito obrigado pela contribuição, estou validando as alterações e gostaria de sugerir:
> 
> * Na pasta pg-dump, deixar  ela apenas com o .gitignore, assim ajuda no dia a dia, evitando fazer o commit da base inicial
> * Criar a documentação em um novo readme apena para o docker, para facilitar o entendimento e as possíilidades de configuração.
> 
> Melhoria: Com essa ambla melhoria, restou apenas a possibilidade de alterar o postgresql.conf na imagem do postgres, consegue ver isso?

Olá @jrbarros,

Realizei as alterações conforme apontadas por você, com esse PR aprovado, facilitará a evolução da infraestrutura como código e até mesmo melhorias na documentação de forma incremental a medida que evoluímos as modificações.

jrbarros commented:
@marcosoliveirasoares94 estou testando as novas alterações. Muito obrigado pela disponibilidade.

marcosoliveirasoares94 commented:
> @marcosoliveirasoares94 estou testando as novas alterações. Muito obrigado pela disponibilidade.

@jrbarros, OK, fique a vontade para reportar qualquer ajuste.

msoares94 committed:
Alterando o locale para en_US

marcosoliveirasoares94 commented:
@jrbarros, @edersoares e @tiagocamargo, existe algum impedimento a ser tratado ainda nesse PR?

jrbarros commented:
@marcosoliveirasoares94, desculpa a demora, estavamos focando em uma atividade bem importante para o futuro do i-educar e acabamos ficando sem tempo. Creio que na próxima semana vamos validar tudo certinho.

edersoares commented on README-DOCKER.md:
Não vejo necessidade de termos este arquivo isolado, tudo deve ficar no readme.

edersoares commented on docker-compose.yml:
```diff
@@ -2,44 +2,60 @@ version: '3'
+    environment:
+      - SERVICE_TAGS=dev
+      - SERVICE_NAME=php
+      - PGPASSWORD=ieducar
+      - PHP_IDE_CONFIG=serverName=i-Educar
+      - XDEBUG_CONFIG=remote_host=127.0.0.1'
```
Remover, são configurações individuais e o arquivo `docker-compose.override.yml` deve ser utilizado.

```suggestion
```

msoares94 replied:
Concordo com esse item!

edersoares commented on docker/nginx/app.conf:
```diff
@@ -1,15 +1,51 @@
     access_log /var/log/nginx/access.log;
 
+    include snippets/params.conf;
+    #include snippets/ssl-params.conf;
```
Eu vejo que deve haver um `include snippets/*.conf` para que individualmente em cada ambiente seja configurado sem impactar no código original.

msoares94 replied:
Muito boa a sua observação @edersoares!

edersoares commented on docker/nginx/app.conf:
```diff
@@ -1,15 +1,51 @@
+    }
+
+    # serve static files directly
+	location ~* \.(jpg|jpeg|gif|css|png|js|ico|html)$ {
```
`html` muitas vezes pode mudar para nunca expirar. Entendo que o TTL deve existir.

msoares94 replied:
@edersoares, nesse caso, recomenda mudar esse trecho `expires max;` por qual valor ?

edersoares commented on docker/nginx/app.conf:
```diff
@@ -1,15 +1,51 @@
+		log_not_found off;
+	}
+
+    # removes trailing slashes (prevents SEO duplicate content issues)
+	if (!-d $request_filename) {
+		rewrite ^/(.+)/$ /$1 permanent;
+	}
+
+    # enforce NO www
+	if ($host ~* ^www\.(.*)) {
+		set $host_without_www $1;
+		rewrite ^/(.*)$ $scheme://$host_without_www/$1 permanent;
+	}
```
Não é necessário.

```suggestion
```

msoares94 replied:
Não é necessário os dois itens ? Ou somente o enforce no www ?

edersoares commented on docker/nginx/app.conf:
```diff
@@ -1,15 +1,51 @@
+	}
+
+    # unless the request is for a valid file (image, js, css, etc.), send to bootstrap
+	if (!-e $request_filename) {
+		rewrite ^/(.*)$ /index.php?/$1 last;
+		break;
+	}
```
Interessante, consegue trazer mais informações sobre?

msoares94 replied:
Encaminhe todas as solicitações para um único script, mais olhando bem a configuração atual o bloco abaixo já resolve isso!

    location / {
        try_files $uri $uri/ /index.php?$query_string;
    }

edersoares commented on docker/nginx/empty.conf:
Arquivo pode ser removido, não há necessidade de existir.

msoares94 replied:
Na minha visão se faz necessário, vamos lá!

Esse arquivo vazio subistitui o `/etc/nginx/conf.d/default.conf` pois ele aponta para o `root /var/www/html;` fazendo com que seja carregado o arquivo `index.html` que vem na instalação default do NGinx e para completar é nesse `default.conf` temos as seguintes configurações como default_server

```
        listen 80 default_server;
        listen [::]:80 default_server;
```

Fazendo com que as requisições na porta 80 seja tratadas nesse arquivo!

Poderíamos até remover tal arquivo `empty.conf` porém deveríamos também fazer a remoção do link simbólico dentro de `/etc/nginx/sites-enabled/` para o `default.conf`

Lembrando, a ideia é facilitar a subida da aplicação i-Educar em ambiente de desenvolvimento não vejo como isso poderia ser ruim pensando em Comunidade!

edersoares commented on docker/nginx/snippets/params.conf:
Arquivo removido, tais configurações devem ser agnósticas e cada usuário deve saber configurá-las individualmente para o seu servidor.

msoares94 replied:
Então, segue a mesma linha de pensamento de facilidade para construção do ambiente de desenvolvimento na mínima configuração viável!

As configurações presentes nesse aquivo facilitam a comunidade do i-Educar subir um ambiente de desenvolvimento mais completo e de fácil customização!

A remoção desse arquivo e a inclusão do treço na documentação mais dificulta a adoção de novos desenvolvedores do que facilita!

edersoares commented on docker/nginx/snippets/ssl-params.conf:
Pode ser removido e uma documentação criada.

msoares94 replied:
Esse caso é igual ao caso do arquivo `params.conf`

E vai além devido ao fato dos navegadores modernos obrigar agora o uso de certificados, nem que para isso seje necessário um certificado auto-assinado, e como a ideia aqui é facilitar a vida das pessoas, resolvi manter isso já realizado!

Sim a documentação seria o complemento, porém a presença de tal configuração certamente mais ajuda do que atrapalha e não impacta em nada em relação a subir essa aplicação em produção, tendo em vista que isso pode ser customizado e deve ser customizado para ambiente de produção!

edersoares commented on docker/nginx/ssl/cert.crt:
Remover.

msoares94 replied:
Discussão desse item nos comentários dos itens `ssl-params.conf` e `params.conf`

edersoares commented on docker/nginx/ssl/cert.key:
Remover.

msoares94 replied:
Discussão desse item nos comentários dos itens `ssl-params.conf` e `params.conf`

edersoares commented on docker/php/8.1/fpm-alpine/php-dev.ini:
Remover. O usuário deve saber customizar uma imagem Docker.

msoares94 replied:
@edersoares, não sou de acordo com essa parte e vou colocar meu ponto de vista e vamos discutir mais a respeito!

Não vejo que a presença desse arquivo seja algo ruim pelo contrário, isso torna a construção do ambiente de DESENVOLVIMENTO mais fácil e mais customizável, pois lembrando que para levar essa aplicação para produção precisa ser personalizado para atender determinadas regras e/ou características e esse arquivo pode facilmente ser substituido por um outro arquivo por exemplo php-prod.ini seja via volume ou processo de compilação de imagem docker.

Lembrando que dentro do Dockerfile ele não está sendo referenciado, ou seja, a imagem será gerada sem ele, ele está presenta para facilitar a referência no docker-compose.override.yml caso algum desenvolvedor da comunidade queira utilizar e queira incrementar mais customizações de desenvolvimento para devolver à comunidade através desse arquivo.

edersoares commented on docker/postgres/.gitkeep:
Remover.

msoares94 replied:
Removido!

edersoares commented on docker/postgres/pg_hba.conf:
Remover, o usuário deve saber configurar individualmente uma imagem docker.

msoares94 replied:
Então, segue a mesma linha de pensamento de facilidade para construção do ambiente de desenvolvimento na mínima configuração viável!

As configurações presentes nesse aquivo facilitam a comunidade do i-Educar subir um ambiente de desenvolvimento mais completo e de fácil customização!

A remoção desse arquivo e a inclusão do treço na documentação mais dificulta a adoção de novos desenvolvedores do que facilita!

edersoares commented on docker/postgres/postgresql.conf:
Remover.

msoares94 replied:
Então, segue a mesma linha de pensamento de facilidade para construção do ambiente de desenvolvimento na mínima configuração viável!

As configurações presentes nesse aquivo facilitam a comunidade do i-Educar subir um ambiente de desenvolvimento mais completo e de fácil customização!

A remoção desse arquivo e a inclusão do treço na documentação mais dificulta a adoção de novos desenvolvedores do que facilita!

edersoares reviewed (commented):
O ambiente de desenvolvimento ou de uso por cada usuário deve ser agnóstico e não haver configurações pré-definidas para itens específicos, como por exemplo certificados, locale e outros.

Se for preciso, o desenvolvedor precisa customizar sua instalação para isso, o projeto tem que fornecer o básico para o funcionamento.

Queremos dar manutenção a nível de código e não de servidor, as configurações específicas para servidores menores funcionam de forma diferente de servidores maiores, então é de responsabilidade do dev/infra saber o que é melhor para cada um.

Obrigado pela contribuição @marcosoliveirasoares94, acredito que levantou excelentes pontos para discutirmos.

msoares94 committed:
Ajuste nas configurações individuais e o arquivo docker-compose.override.yml deve ser utilizado.

msoares94 committed:
Delete README-DOCKER.md

msoares94 committed:
Mantendo o README.md oficial do projeto!

msoares94 committed:
Modificando include

Incluindo  um include snippets/*.conf para que individualmente em cada ambiente seja configurado sem impactar no código original.

msoares94 committed:
Delete .gitkeep

marcosoliveirasoares94 commented:
> @marcosoliveirasoares94, desculpa a demora, estavamos focando em uma atividade bem importante para o futuro do i-educar e acabamos ficando sem tempo. Creio que na próxima semana vamos validar tudo certinho.

Obrigado pelo retorno @jrbarros!

Realizei comentários nos apontamentos do @edersoares, caso queira fique a vontade para discutirmos conosco!
